src/BinSkim.Rules/DwarfRules/BA3003.EnableStackProtector.cs

```diff
@@ -64,70 +71,98 @@ public override AnalysisApplicability CanAnalyzeDwarf(IDwarfBinary target, Sarif
                 {
-                    if (ElfUtility.GetDwarfCommandLineType(info.CommandLine) != DwarfCommandLineType.Gcc)
-                    {
-                        continue;
-                    }
+                    failed = true;
+                }
+                else
+                {
+                    string[] paramToCheck = { "--param=ssp-buffer-size=" };
+                    string paramValue = string.Empty;
+                    bool found = GetOptionValue(info.CommandLine, paramToCheck, OrderOfPrecedence.FirstWins, ref paramValue);
 
-                    bool failed = false;
-                    if ((!info.CommandLine.Contains("-fstack-protector-all", StringComparison.OrdinalIgnoreCase)
-                        && !info.CommandLine.Contains("-fstack-protector-strong", StringComparison.OrdinalIgnoreCase))
-                        || info.CommandLine.Contains("-fno-stack-protector", StringComparison.OrdinalIgnoreCase))
-                    {
-                        failed = true;
-                    }
-                    else
+                    if (found && !string.IsNullOrWhiteSpace(paramValue))
                     {
-                        string[] paramToCheck = { "--param=ssp-buffer-size=" };
-                        string paramValue = string.Empty;
-                        bool found = GetOptionValue(info.CommandLine, paramToCheck, OrderOfPrecedence.FirstWins, ref paramValue);
-
-                        if (found && !string.IsNullOrWhiteSpace(paramValue))
+                        if (int.TryParse(paramValue, out int bufferSize))
                         {
-                            if (int.TryParse(paramValue, out int bufferSize))
+                            if (bufferSize > 4)
                             {
-                                if (bufferSize > 4)
-                                {
-                                    failed = true;
-                                }
+                                failed = true;
                             }
                         }
                     }
+                }
+
+                if (failed)
+                {
+                    failedList.Add(info);
+                }
+            }
+
+            return !failedList.Any();
+        }
 
-                    if (failed)
+        bool analyzeSymbols(ElfBinary binary)
+        {
+            foreach (ISection section in binary.ELF.Sections)
+            {
+                if (section.Type == SectionType.DynamicSymbolTable)
+                {
+                    var symbols = section as SymbolTable<ulong>;
+                    foreach (SymbolEntry<ulong> symbol in symbols.Entries)
                     {
-                        failedList.Add(info);
+                        if (symbol.Name == "__stack_chk_fail" || symbol.Name == "stack_chk_guard" || symbol.Name == "__intel_security_cookie")
```

Comment:
>if (symbol.Name == "__stack_chk_fail" || symbol.Name == "stack_chk_guard" || symbol.Name == "__intel_security_cookie") [](http://example.com/codeflow?start=24&length=118)

Sorry not very familiar, can you provide webpage/doc about these 3 are the ones that is should be at least one present and also can indicates Stack Protector enabled with proper level.

Reply:
I'm not aware of any documentation that concretely describes how stack cookies are implemented at this level, unfortunately. This is based off checksec's implementation here: https://github.com/slimm609/checksec.sh/blob/main/checksec#L679. From my manual analysis, I've always seen __stack_chk_fail from libc being used to trigger a stack check failure. This method alone cannot detect the stack protector level used, as we're effectively just checking for the presence of *some* stack check in the entire binary. That's why I'm treating this as a fallback. I did just notice that I typo'd stack_chk_guard; it needs to be __stack_chk_guard. I'll fix that.